core/pkg/evaluator/json.go

```diff
@@ -119,15 +119,16 @@ func (je *JSON) SetState(payload sync.DataSync) (map[string]interface{}, bool, e
 	var events map[string]interface{}
 	var reSync bool
 
+	// TODO: We do not handle metadata in ADD/UPDATE operations. These are only relevant for grpc sync implementations.
 	switch payload.Type {
 	case sync.ALL:
-		events, reSync = je.store.Merge(je.Logger, payload.Source, payload.Selector, newFlags.Flags)
+		events, reSync = je.store.Merge(je.Logger, payload.Source, payload.Selector, definition.Flags, definition.Metadata)
 	case sync.ADD:
-		events = je.store.Add(je.Logger, payload.Source, payload.Selector, newFlags.Flags)
+		events = je.store.Add(je.Logger, payload.Source, payload.Selector, definition.Flags)
 	case sync.UPDATE:
-		events = je.store.Update(je.Logger, payload.Source, payload.Selector, newFlags.Flags)
+		events = je.store.Update(je.Logger, payload.Source, payload.Selector, definition.Flags)
 	case sync.DELETE:
-		events = je.store.DeleteFlags(je.Logger, payload.Source, newFlags.Flags)
+		events = je.store.DeleteFlags(je.Logger, payload.Source, definition.Flags)
```

Comment:
ADD/UPDATE are only relevant to hypothetical gRPC server implementations, and were a premature optimization to allow gRPC servers to tell flagd "just add these flags, or just change this flag" instead of sending a whole payload. I have never seen this implemented, and it's not relevant to any of the other sync types.

I think we should cease to enhance them and deprecate them in the protobuf.